compiler/rustc_metadata/src/locator.rs

```diff
@@ -394,46 +398,97 @@ impl<'a> CrateLocator<'a> {
         // of the crate id (path/name/id).
         //
         // The goal of this step is to look at as little metadata as possible.
-        self.filesearch.search(|spf, kind| {
-            let file = match &spf.file_name_str {
-                None => return FileDoesntMatch,
-                Some(file) => file,
+        for search_path in self.filesearch.search_paths() {
+            debug!("searching {}", search_path.dir.display());
+
+            // njn: factor out more similarities?
+
+            /*
+            let mut s = String::new();
+            s.push_str(&format!("find {} in {}\n", rlib_prefix, search_path.dir.display()));
+            for rlib in rlibs {
+                s.push_str(&format!("- {}\n", rlib.file_name_str));
+            }
+            s.push_str(&format!("linear {:?}\n", linear_matches));
+            s.push_str(&format!("binary {:?}\n", binary_matches));
+            eprint!("***\n{}***\n", s);
+            */
+
+            // njn: rename
+            // njn: kind isn't needed as an arg
+            let mut add_candidate = |table: &mut FxHashMap<_, _>, path: &PathBuf| {
+                let path = fs::canonicalize(&path).unwrap_or_else(|_| path.clone());
+                if !seen_paths.contains(&path) {
+                    seen_paths.insert(path.clone());
+                    table.insert(path, search_path.kind);
+                };
             };
-            let (hash, found_kind) = if file.starts_with(&rlib_prefix) && file.ends_with(".rlib") {
-                (&file[(rlib_prefix.len())..(file.len() - ".rlib".len())], CrateFlavor::Rlib)
-            } else if file.starts_with(&rlib_prefix) && file.ends_with(".rmeta") {
-                (&file[(rlib_prefix.len())..(file.len() - ".rmeta".len())], CrateFlavor::Rmeta)
-            } else if file.starts_with(&dylib_prefix) && file.ends_with(&self.target.dll_suffix) {
-                (
-                    &file[(dylib_prefix.len())..(file.len() - self.target.dll_suffix.len())],
-                    CrateFlavor::Dylib,
-                )
-            } else {
-                if file.starts_with(&staticlib_prefix)
-                    && file.ends_with(&self.target.staticlib_suffix)
+
+            // .rlib files
+            let rlibs = &search_path.rlibs;
+            // njn: explain this
+            let i = rlibs.partition_point(|spf| spf.file_name_str < rlib_prefix);
+            let mut j = i;
+            while let Some(true) =
+                rlibs.get(j).map(|spf| spf.file_name_str.starts_with(&rlib_prefix))
+            {
+                j += 1;
+            }
+
+            // Process the matches.
+            for spf in &rlibs[i..j] {
+                let name = &spf.file_name_str;
+                assert!(name.ends_with(".rlib"));
+                let hash = &name[(rlib_prefix.len())..(name.len() - ".rlib".len())];
```

Comment:
This can use `trim_end_matches`.

Reply:
This code is trimming text from the start and the end of the filename, so a slice seems better to me...